internal/pipeline/step.go

```diff
@@ -0,0 +1,294 @@
+package pipeline
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/buildkite/agent/v3/internal/yamltojson"
+	"gopkg.in/yaml.v3"
+)
+
+var _ signedFielder = (*CommandStep)(nil)
+
+// Step models a step in the pipeline. It will be a pointer to one of:
+// - CommandStep
+// - WaitStep
+// - InputStep
+// - TriggerStep
+// - GroupStep
+type Step interface {
+	stepTag() // allow only the step types below
+
+	selfInterpolater
+}
+
+// CommandStep models a command step.
+//
+// Standard caveats apply - see the package comment.
+type CommandStep struct {
+	Command   string     `yaml:"command"`
+	Label     string     `yaml:"label,omitempty"`
+	Plugins   Plugins    `yaml:"plugins,omitempty"`
+	Signature *Signature `yaml:"signature,omitempty"`
+
+	// RemainingFields stores any other top-level mapping items so they at least
+	// survive an unmarshal-marshal round-trip.
+	RemainingFields map[string]any `yaml:",inline"`
+}
+
+// MarshalJSON marshals a pipeline to JSON. Special handling is needed because
+// yaml.v3 has "inline" but encoding/json has no concept of it.
+func (c *CommandStep) MarshalJSON() ([]byte, error) {
+	out := map[string]any{
+		"command": c.Command,
+	}
+	if c.Label != "" {
+		out["label"] = c.Label
+	}
+	if len(c.Plugins) > 0 {
+		out["plugins"] = c.Plugins
+	}
+	if c.Signature != nil {
+		out["signature"] = c.Signature
+	}
+	for k, v := range c.RemainingFields {
+		if v != nil {
+			out[k] = v
+		}
+	}
+	return json.Marshal(out)
+}
+
+func (c *CommandStep) UnmarshalYAML(n *yaml.Node) error {
+	// Unmarshal into this secret type, then normalise.
+	var full struct {
+		Command   string     `yaml:"command,omitempty"`
+		Commands  []string   `yaml:"commands,omitempty"`
+		Label     string     `yaml:"label,omitempty"`
+		Plugins   Plugins    `yaml:"plugins,omitempty"`
+		Signature *Signature `yaml:"signature,omitempty"`
+
+		RemainingFields map[string]any `yaml:",inline"`
+	}
+	if err := n.Decode(&full); err != nil {
+		return err
+	}
+
+	// Normalise into Command.
+	c.Command = full.Command
+	if c.Command == "" {
+		c.Command = strings.Join(full.Commands, "\n")
+	}
```

Comment:
i wonder if we should think about doing this the other way around - ie have `commands` be the "canonical" representation of commands. 

my reasoning here is that the `command` -> `commands` normalisation eg:
```yaml
command: >
  echo hello
  echo there
```
to 
```yaml
commands: ["echo hello", "echo there"]
```
feels like it retains more specificity than the other way around.

Reply:
Thinking through this...

We upload `command` or `commands` but get back `BUILDKITE_COMMAND`. 

It should be easy to ensure consistency when computing signatures for `command` (1 string) and `BUILDKITE_COMMAND` (1 string). 

If we split `command` into `commands` (N strings), then we need to split `BUILDKITE_COMMAND` into the same N strings for the signatures to be consistent. If the user supplies `commands` and an element of `commands` contains multiple lines then this won't be consistent - we'd need to normalise by splitting each element of `commands`. Or we'd need to `strings.Join(commands, "\n")` solely for the benefit of computing signatures.

Reply:
yep, that makes total sense to me 👍 very happy to leave as is